flytekit/common/tasks/notebook.py

```diff
@@ -0,0 +1,542 @@
+from flytekit.common.tasks import task as base_tasks
+from flytekit.engines import loader as _engine_loader
+
+type_map = {
+    int: _p.Integer,
+    bool: _p.Boolean,
+    float: _p.Float,
+    str: _p.String,
+    _datetime.datetime: _p.Datetime,
+    _datetime.timedelta: _p.Timedelta,
+}
+
+OUTPUT_NOTEBOOK = 'output_notebook'
+
+def convert_outputs(outputs=None):
+    """
+    outputs: dict
+    Records the output
+    """
+    if outputs is None:
+        return _packer({}, {})
+    tm = {}
+    for k, v in _six.iteritems(outputs):
+        t = type(v)
+        if t not in type_map:
+            raise ValueError(
+                "Currently only primitive types {} are supported for recording from notebook".format(type_map))
+        tm[k] = type_map[t]
+    return _packer(outputs, tm)
+
+
+def record_outputs(outputs=None):
+    p = convert_outputs(outputs)
+    return p.to_flyte_idl()
+
+# TODO: Support Client Mode
+def get_spark_context(spark_conf):
+    # We run in cluster-mode in Flyte.
+    # Ref https://github.com/lyft/flyteplugins/blob/master/go/tasks/v1/flytek8s/k8s_resource_adds.go#L46
+    if "FLYTE_INTERNAL_EXECUTION_ID" in _os.environ:
+        return SparkContext()
+
+    # Add system spark-conf for local/notebook based execution.
+    spark_conf.add(("spark.master", "local"))
+    conf = SparkConf().setAll(spark_conf)
+    return SparkContext(conf=conf)
+
+def python_notebook(
+        notebook_path='',
+        inputs=None,
+        outputs=None,
+        cache_version='',
+        retries=0,
+        deprecated='',
+        storage_request=None,
+        cpu_request=None,
+        gpu_request=None,
+        memory_request=None,
+        storage_limit=None,
+        cpu_limit=None,
+        gpu_limit=None,
+        memory_limit=None,
+        cache=False,
+        timeout=None,
+        environment=None,
+        cls=None,
+
+):
+    """
+    Decorator to create a Python Notebook Task definition.  This task will run as a single unit of work on the platform.
+
+    :rtype: flytekit.common.tasks.sdk_runnable.SdkNotebookTask
+    """
+    return SdkNotebookTask(
+            notebook_path=notebook_path,
+            inputs=inputs,
+            outputs=outputs,
+            task_type=_constants.SdkTaskType.PYTHON_TASK,
+            discovery_version=cache_version,
+            retries=retries,
+            deprecated=deprecated,
+            storage_request=storage_request,
+            cpu_request=cpu_request,
+            gpu_request=gpu_request,
+            memory_request=memory_request,
+            storage_limit=storage_limit,
+            cpu_limit=cpu_limit,
+            gpu_limit=gpu_limit,
+            memory_limit=memory_limit,
+            discoverable=cache,
+            timeout=timeout or _datetime.timedelta(seconds=0),
+            environment=environment,
+            custom={})
+
+
+class SdkNotebookTask(
+        _six.with_metaclass(sdk_bases.ExtendedSdkType, base_tasks.SdkTask)):
```

Comment:
Can we make this derive from SdkRunnableTask?  If not, we should massage SdkRunnableTask to allow this use-case to work.

Reply:
It is quite a bit of work and would involve most of SdkRunnableTask methods to be overridden. leaving as-is based on our discussion